### PR TITLE
[Grappler]Fuse Conv2D + BiasAdd + Relu on GPU.

### PR DESCRIPTION
**Pull request by mpjlu**
Author: Dr. Peng Meng from Tencent Cloud TI-ML team.
Grappler:remapper Conv2D + BiasAdd + Relu OP fuse doesn't work on GPU.
To fuse these three OP,  FindContractionWithBiasAndActivation needs to know the InputProperties:
image
TF master code  doesn't call graph_properties.InferStatically() before calling FindContractionWithBiasAndActivation,  FindContractionWithBiasAndActivation always return false, so Conv2D + BiasAdd + Relu cannot be fused.

This PR fixes this bug. After this fix, the WDSR model inference latency improves 20% on Nvidia T4 GPU.

### CODE REVIEW

**Comment by rmlarsen** [comment]
Thanks a lot for the fix! Could you please update the unit test to make sure we do not regress on this again?

**Comment by mpjlu** [comment]
@rmlarsen Thanks for your review. I have added the unit test.

**Review by ezhulenev — DISMISSED** [review]

**Review by ezhulenev — DISMISSED** [review]

**Comment by gbaned** [comment]
@mpjlu Can you please resolve conflicts? Thanks!

**Comment by mpjlu** [comment]
> @mpjlu Can you please resolve conflicts? Thanks!

Resolved, Thanks!

**Comment by mpjlu** [comment]
@rmlarsen @ezhulenev could you please review this PR again, thanks.

**Review by ezhulenev — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Review by rthadur — DISMISSED** [review]

File: tensorflow/core/grappler/optimizers/remapper_test.cc [hunk 1]
```diff
@@ -449,6 +449,76 @@ TEST_F(RemapperTest, FuseMatMulWithBias) {
   test::ExpectTensorNear<float>(tensors[0], tensors_expected[0], 1e-6);
 }
 
+TEST_F(RemapperTest, FuseConv2DWithBiasAndActivationOnGPU) {
+  using ::tensorflow::ops::Placeholder;
```
Comment (ezhulenev) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:453]: This test has to be skipped if there are no GPUs available.

**Review by ezhulenev — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/remapper_test.cc (hunk 1 above)
Comment (mpjlu) [reply tensorflow/core/grappler/optimizers/remapper_test.cc:453]: > This test has to be skipped if there are no GPUs available.

Done, thanks!

**Review by ezhulenev — DISMISSED** [review]

File: tensorflow/core/grappler/optimizers/remapper_test.cc [hunk 2]
```diff
@@ -449,6 +449,79 @@ TEST_F(RemapperTest, FuseMatMulWithBias) {
...
+  auto input_t = GenerateRandomTensor<DT_FLOAT>({8, 32, 32, 3});
+  auto filter_t = GenerateRandomTensor<DT_FLOAT>({3, 3, 3, 128});
+  auto bias_t = GenerateRandomTensor<DT_FLOAT>({128});
+
+  GrapplerItem item;
+  item.fetch = {"fetch"};
+  item.feed = {{"input", input_t}, {"filter", filter_t}, {"bias", bias_t}};
+  TF_ASSERT_OK(s.ToGraphDef(&item.graph));
+
+  // Place all nodes on GPU.
```
Comment (rthadur) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:488]: One of the tests are still crashing when run in a config where no GPU device is available. Please remove lines 487-490.

File: tensorflow/core/grappler/optimizers/remapper_test.cc [hunk 3]
```diff
@@ -449,6 +449,79 @@ TEST_F(RemapperTest, FuseMatMulWithBias) {
...
+  for (int i = 0; i < item.graph.node_size(); ++i) {
+    item.graph.mutable_node(i)->set_device("/device:GPU:0");
+  }
+
+  Remapper optimizer(RewriterConfig::AGGRESSIVE);  // trust placeholders shape
```
Comment (rthadur) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:493]: please delete this line.

**Review by rthadur — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/remapper_test.cc (hunk 2 above)
Comment (mpjlu) [reply tensorflow/core/grappler/optimizers/remapper_test.cc]: done, thanks!

**Review by ezhulenev — APPROVED** [review]

File: tensorflow/core/grappler/optimizers/remapper_test.cc [hunk 4]
```diff
@@ -449,6 +449,74 @@ TEST_F(RemapperTest, FuseMatMulWithBias) {
...
+  TF_ASSERT_OK(s.ToGraphDef(&item.graph));
+
+  Remapper optimizer(RewriterConfig::AGGRESSIVE);  // trust placeholders shape
+  //Remapper optimizer(RewriterConfig::ON);
+  GraphDef output;
+  TF_ASSERT_OK(optimizer.Optimize(nullptr, item, &output));
+
+  int found = 0;
+  for (const NodeDef& node : output.node()) {
+    if (node.name() == "activation") {
```
Comment (rthadur) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:495]: `tensorflow/core/grappler/optimizers/remapper_test.cc:495
Expected equality of these values:
  node.op()
    Which is: "Relu"
  "_FusedConv2D"
Stack trace:`

File: tensorflow/core/grappler/optimizers/remapper_test.cc [hunk 5]
```diff
@@ -449,6 +449,74 @@ TEST_F(RemapperTest, FuseMatMulWithBias) {
...
+      EXPECT_EQ(node.op(), "_FusedConv2D");
```
Comment (rthadur) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:496]: `tensorflow/core/grappler/optimizers/remapper_test.cc:496
Expected: (node.input_size()) >= (3), actual: 1 vs 3`

**Review by rthadur — CHANGES_REQUESTED** [review]
@mpjlu can you please make below changes.

File: tensorflow/core/grappler/optimizers/remapper_test.cc (hunk 3 above)
Comment (rthadur) [reply tensorflow/core/grappler/optimizers/remapper_test.cc:488]: Please delete this line

**Review by rthadur — CHANGES_REQUESTED** [review]
Please include lines below again as this was required for GPU, sorry for the confusion
`// Place all nodes on GPU.
  for (int i = 0; i < item.graph.node_size(); ++i) {
    item.graph.mutable_node(i)->set_device("/device:GPU:0");
  }`

**Pull request merged**